ignite/contrib/checkpoint.py

```diff
@@ -0,0 +1,131 @@
+        and return a score (float / int). Objects with highest scores will be retained.
+    n_saved : int, optional
+        Number of objects that should be kept on disk. Older files will be removed.
+    atomic : bool, optional
+        If True, objects are serialized to a temporary file, and then moved to final destination, so that
+        files are guaranteed to not be damaged (for example if exception occures during saving).
+    require_empty : bool, optional
+        If True, will raise exception if there are any files starting with `filename_prefix` in the directory 'dirname'
+    create_dir : bool, optional
+        If True, will create directory 'dirname' if it doesnt exist.
+    exist_ok : bool, optional
+        Passed to 'os.makedirs' call. Ignored if 'create_dir' is False.
+    """
+
+    def __init__(self, dirname, filename_prefix,
+                 iteration_interval=None, epoch_interval=None,
+                 score_function=None,
+                 n_saved=1,
+                 atomic=True, require_empty=True,
+                 create_dir=True, exist_ok=True):
+
+        self._dirname            = dirname
+        self._fname_prefix       = filename_prefix
+        self._n_saved            = n_saved
+        self._iteration_interval = iteration_interval
+        self._epoch_interval     = epoch_interval
+        self._score_function     = score_function
+        self._atomic             = atomic
+
+        self._item_T  = namedtuple('_item_T', ('priority', 'data'))
+        self._saved   = []
+
+        if not (iteration_interval or epoch_interval or score_function):
+            raise ValueError("One of 'iteration_interval', 'epoch_interval' or "
+                             "'score_function' arguments must be provided.")
+
+        if create_dir:
+            os.makedirs(dirname, exist_ok=exist_ok)
+
+        if require_empty:
+            n_matched = [fname
+                         for fname in os.listdir(dirname)
+                         if fname.startswith(self._fname_prefix)]
+
+            if len(n_matched) > 0:
+                raise ValueError("Files prefixed with {} are already present "
+                                 "in the directory {}. If you want to use this "
+                                 "directory anyway, pass require_empty=False. "
+                                 "".format(filename_prefix, dirname))
+
+    def _save(self, obj, path):
+        if not self._atomic:
+            torch.save(obj, path)
+
+        else:
+            tmp = tempfile.NamedTemporaryFile(delete=False)
+
+            try:
+                torch.save(obj, tmp.file)
+
+            except BaseException:
+                tmp.close()
+                os.remove(tmp.name)
+                raise
+
+            else:
+                tmp.close()
+                os.rename(tmp.name, path)
+
+    def __call__(self, engine, **kwargs):
+        if len(kwargs) == 0:
+            raise RuntimeError("No objects to checkpoint found.")
```

Comment:
Does the model need to be a `kwarg`? can it not just be an arg (which can also be a list) and then you get this check for free?

Reply:
I'm not sure how we should name the files where object will be saved, so I decided to write them to `{prefix}_{name}_{index}.pth`, where the `name` is taken from `kwargs`. 

This should be useful if you want to save multiple objects into seperate files, e.g. `encoder` and `decoder`. 

WDYT?

Reply:
maybe a dict of {name : model} - just so that it's explicit in the docstring?

Reply:
Ok, I'll also add an example in the docstring.